compiler/hash-typecheck/src/traverse/mod.rs

```diff
@@ -299,22 +301,36 @@ impl<'gs, 'ls, 'cd, 'src> visitor::AstVisitor for TcVisitor<'gs, 'ls, 'cd, 'src>
         node: hash_ast::ast::AstNodeRef<hash_ast::ast::DerefExpr>,
     ) -> Result<Self::DerefExprRet, Self::Error> {
         let walk::DerefExpr(inner) = walk::walk_deref_expr(self, ctx, node)?;
+        let inner_ty = self.typer().ty_of_term(inner)?;
+
+        let ref_ty = self.core_defs().reference_ty_fn;
+        let ap_ref_ty = {
+            let builder = self.builder();
+
+            builder.create_app_ty_fn_term(
+                ref_ty,
+                builder.create_args(
+                    [builder.create_arg("T", builder.create_unresolved_term())],
+                    ParamOrigin::TyFn,
+                ),
+            )
+        };
 
         // Attempt to unify this with a `Ref<T>` to see if the `inner_ty` can
         // be dereferenced.
         // @@Todo: deal with `RefMut<T>` and raw references...
-        let dummy = self.core_defs().reference_ty_fn;
+        let result = self.unifier().unify_terms(inner_ty, ap_ref_ty)?;
 
-        // Unify the types, get the substitution result and use that as the return from
-        // the `inner_ty`
-        let result = self.unifier().unify_terms(inner, dummy)?;
+        // get the substitution result and use that as the return from the `inner_ty`
         let inner_ty = result.pairs().next().unwrap().1;
```

Comment:
This needs to be attempted both with `Ref` and `RefMut`. If both fail, it should error.